File: sdk/android/captcha/src/main/java/com/engagelab/captcha/EngagelabCaptcha.kt
package com.engagelab.captcha

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume

/**
 * Main entry point for the EngageLab CAPTCHA SDK.
 *
 * Usage:
 * ```kotlin
 * val config = CaptchaConfig(siteKey = "your-site-key", scene = SceneType.LOGIN)
 * val captcha = EngagelabCaptcha(config)
 * val result = captcha.execute(context)
 * when (result) {
 *     is CaptchaResult.Passed -> // send result.token to your backend
 *     is CaptchaResult.ChallengeRequired -> // present challenge UI
 *     is CaptchaResult.Denied -> // block the action
 *     is CaptchaResult.Error -> // handle result.exception
 * }
 * ```
 */
class EngagelabCaptcha(
    val config: CaptchaConfig
) {

    // ----------------------------------------------------------------
    // Errors
    // ----------------------------------------------------------------

    class CaptchaException(message: String, cause: Throwable? = null) : Exception(message, cause)

    // ----------------------------------------------------------------
    // Properties
    // ----------------------------------------------------------------

    val behaviorTracker = BehaviorTracker()

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .build()

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    // ----------------------------------------------------------------
    // Execute
    // ----------------------------------------------------------------

    /**
     * Run the full CAPTCHA flow: precheck -> optional challenge -> result.
     *
     * @param context Android context used for device info collection.
     */
    suspend fun execute(context: Context): CaptchaResult {
        return try {
            // Validate config
            if (config.siteKey.isBlank()) {
                return CaptchaResult.Error(CaptchaException("siteKey must not be empty"))
            }

            // Step 1: Collect signals
            val deviceInfo = DeviceInfoCollector.collect(context)
            val behavior = behaviorTracker.behaviorData()

            // Step 2: Precheck
            val precheckResponse = precheck(deviceInfo, behavior)

            // Step 3: Handle action
            when (precheckResponse.action) {
                ChallengeAction.PASS -> CaptchaResult.Passed(precheckResponse.token)
                ChallengeAction.INVISIBLE -> CaptchaResult.Passed(precheckResponse.token)
                ChallengeAction.CHALLENGE -> {
                    val challengeId = precheckResponse.challengeId
                    val challengeType = precheckResponse.challengeType
                    if (challengeId != null && challengeType != null) {
                        CaptchaResult.ChallengeRequired(challengeId, challengeType)
                    } else {
                        CaptchaResult.Error(
                            CaptchaException("Challenge required but no challenge_id or challenge_type returned")
                        )
                    }
                }
                ChallengeAction.DENY -> CaptchaResult.Denied
            }
        } catch (e: Exception) {
            CaptchaResult.Error(e)
        }
    }

    // ----------------------------------------------------------------
    // Verify
    // ----------------------------------------------------------------

    /**
     * Submit the user's challenge answer and get a verification result.
     */
    suspend fun verify(challengeId: String, answer: String): CaptchaResult {
        return try {
            val response = submitChallengeAnswer(challengeId, answer)
            if (response.success && response.token != null) {
                CaptchaResult.Passed(response.token)
            } else {
                val message = response.errorMessage ?: "Verification failed"
                CaptchaResult.Error(CaptchaException(message))
            }
        } catch (e: Exception) {
            CaptchaResult.Error(e)
        }
    }

    // ----------------------------------------------------------------
    // Challenge WebView
    // ----------------------------------------------------------------

    /**
     * Present a WebView to render the challenge and return the answer.
     * Call this when [execute] returns [CaptchaResult.ChallengeRequired].
     *
     * @param activity The activity to host the WebView.
     * @param challengeId The challenge ID from the precheck response.
     * @param challengeType The challenge type from the precheck response.
     */
    suspend fun presentChallenge(
        activity: Activity,
        challengeId: String,
        challengeType: String
    ): CaptchaResult {
        return try {
            // Fetch render data
            val renderResponse = fetchChallengeRender(challengeId)

            // Show WebView and wait for answer
            val answer = showChallengeWebView(activity, renderResponse, challengeId)
                ?: return CaptchaResult.Error(CaptchaException("Challenge cancelled"))

            // Verify
            verify(challengeId, answer)
        } catch (e: Exception) {
            CaptchaResult.Error(e)
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private suspend fun showChallengeWebView(
        activity: Activity,
        renderResponse: ChallengeRenderResponse,
        challengeId: String
    ): String? = suspendCancellableCoroutine { cont ->
        activity.runOnUiThread {
            val container = FrameLayout(activity).apply {
                layoutParams = ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT
                )
            }

            val webView = WebView(activity).apply {
                layoutParams = ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT
                )
                settings.javaScriptEnabled = true
                settings.domStorageEnabled = true
                webChromeClient = WebChromeClient()
                webViewClient = WebViewClient()
            }

            // Add JS bridge
            webView.addJavascriptInterface(object {
                @JavascriptInterface
                fun submit(answer: String) {
                    activity.runOnUiThread {
                        (container.parent as? ViewGroup)?.removeView(container)
                        if (cont.isActive) cont.resume(answer)
                    }
                }

                @JavascriptInterface
                fun cancel() {
                    activity.runOnUiThread {
                        (container.parent as? ViewGroup)?.removeView(container)
                        if (cont.isActive) cont.resume(null)
                    }
                }
            }, "EngagelabCaptcha")

            container.addView(webView)

            // Add to the activity's root view
            val rootView = activity.findViewById<ViewGroup>(android.R.id.content)
            rootView.addView(container)

            // Build URL
            val uri = Uri.parse(renderResponse.renderUrl).buildUpon()
                .appendQueryParameter("challenge_id", challengeId)
                .appendQueryParameter("theme", config.theme.value)
                .apply { config.lang?.let { appendQueryParameter("lang", it) } }
                .appendQueryParameter("platform", "android")
                .build()

            webView.loadUrl(uri.toString())

            cont.invokeOnCancellation {
                activity.runOnUiThread {
                    (container.parent as? ViewGroup)?.removeView(container)
                }
            }
        }
    }

    // ----------------------------------------------------------------
    // Network
    // ----------------------------------------------------------------

    private suspend fun precheck(
        deviceInfo: Map<String, String>,
        behaviorData: Map<String, String>
    ): PrecheckResponse {
        val body = PrecheckRequest(
            siteKey = config.siteKey,
            scene = config.scene.value,
            deviceInfo = deviceInfo,
            behaviorData = behaviorData,
            lang = config.lang,
            theme = config.theme.value
        )
        return post("/precheck", body)
    }

    private suspend fun fetchChallengeRender(challengeId: String): ChallengeRenderResponse {
        val url = buildUrl("/challenge/render", mapOf(
            "challenge_id" to challengeId,
            "site_key" to config.siteKey
        ))
        return get(url)
    }

    private suspend fun submitChallengeAnswer(
        challengeId: String,
        answer: String
    ): ChallengeVerifyResponse {
        val body = ChallengeVerifyRequest(
            challengeId = challengeId,
            answer = answer,
            token = "" // Token will be set by the server from the session
        )
        return post("/challenge/verify", body)
    }

    // ----------------------------------------------------------------
    // HTTP Helpers
    // ----------------------------------------------------------------

    private fun buildUrl(path: String, query: Map<String, String>? = null): String {
        val base = config.apiBase.trimEnd('/') + path
        if (query.isNullOrEmpty()) return base
        val uri = Uri.parse(base).buildUpon()
        query.forEach { (k, v) -> uri.appendQueryParameter(k, v) }
        return uri.build().toString()
    }

    private fun userAgent(): String {
        return "EngagelabCaptcha-Android/1.0.0"
    }

    private suspend inline fun <reified T : Any> post(path: String, body: Any): T {
        return withContext(Dispatchers.IO) {
            val url = buildUrl(path)
            val jsonBody = json.encodeToString(
                kotlinx.serialization.serializer(body::class.java),
                body
            )
            val request = Request.Builder()
                .url(url)
                .post(jsonBody.toRequestBody(jsonMediaType))
                .header("Content-Type", "application/json")
                .header("User-Agent", userAgent())
                .build()

            executeRequest(request)
        }
    }

    private suspend inline fun <reified T : Any> get(url: String): T {
        return withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url(url)
                .get()
                .header("Accept", "application/json")
                .header("User-Agent", userAgent())
                .build()

            executeRequest(request)
        }
    }

    private inline fun <reified T : Any> executeRequest(request: Request): T {
        val response = try {
            httpClient.newCall(request).execute()
        } catch (e: IOException) {
            throw CaptchaException("Network error: ${e.message}", e)
        }

        val responseBody = response.body?.string()
            ?: throw CaptchaException("Empty response body, status=${response.code}")

        if (!response.isSuccessful) {
            throw CaptchaException("Server error ${response.code}: $responseBody")
        }

        return try {
            json.decodeFromString<T>(responseBody)
        } catch (e: Exception) {
            throw CaptchaException("Decoding error: ${e.message}", e)
        }
    }
}

/**
 * Optional: Standalone Activity that hosts the challenge WebView.
 * Can be launched via Intent when a full-screen challenge experience is desired.
 */
class ChallengeActivity : AppCompatActivity() {

    companion object {
        const val EXTRA_RENDER_URL = "render_url"
        const val EXTRA_CHALLENGE_ID = "challenge_id"
        const val EXTRA_THEME = "theme"
        const val EXTRA_LANG = "lang"
        const val RESULT_ANSWER = "answer"

        internal val pendingResult = CompletableDeferred<String?>()
    }

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val renderUrl = intent.getStringExtra(EXTRA_RENDER_URL) ?: run {
            pendingResult.complete(null)
            finish()
            return
        }
        val challengeId = intent.getStringExtra(EXTRA_CHALLENGE_ID) ?: ""
        val theme = intent.getStringExtra(EXTRA_THEME) ?: "light"
        val lang = intent.getStringExtra(EXTRA_LANG)

        val webView = WebView(this).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
            settings.javaScriptEnabled = true
            settings.domStorageEnabled = true
            webChromeClient = WebChromeClient()
            webViewClient = WebViewClient()
        }

        webView.addJavascriptInterface(object {
            @JavascriptInterface
            fun submit(answer: String) {
                pendingResult.complete(answer)
                runOnUiThread { finish() }
            }

            @JavascriptInterface
            fun cancel() {
                pendingResult.complete(null)
                runOnUiThread { finish() }
            }
        }, "EngagelabCaptcha")

        setContentView(webView)

        val uri = Uri.parse(renderUrl).buildUpon()
            .appendQueryParameter("challenge_id", challengeId)
            .appendQueryParameter("theme", theme)
            .apply { lang?.let { appendQueryParameter("lang", it) } }
            .appendQueryParameter("platform", "android")
            .build()

        webView.loadUrl(uri.toString())
    }

    override fun onBackPressed() {
        super.onBackPressed()
        pendingResult.complete(null)
    }
}

File: sdk/ios/Sources/EngagelabCaptcha/EngagelabCaptcha.swift
import Foundation
#if canImport(UIKit)
import UIKit
import WebKit
#endif

/// Main entry point for the EngageLab CAPTCHA SDK.
///
/// Usage:
/// ```swift
/// let config = CaptchaConfig(siteKey: "your-site-key", scene: .login)
/// let captcha = EngagelabCaptcha(config: config)
/// let result = await captcha.execute()
/// switch result {
/// case .passed(let token):
///     // send token to your backend
/// case .challengeRequired(let challengeID, let type):
///     // present challenge UI
/// case .denied:
///     // block the action
/// case .error(let err):
///     // handle error
/// }
/// ```
public final class EngagelabCaptcha: @unchecked Sendable {

    // MARK: - Errors

    public enum CaptchaError: Error, LocalizedError {
        case invalidConfig(String)
        case networkError(Error)
        case serverError(statusCode: Int, message: String)
        case decodingError(Error)
        case challengeTimeout
        case challengeFailed(String)
        case webViewUnavailable
        case cancelled

        public var errorDescription: String? {
            switch self {
            case .invalidConfig(let msg): return "Invalid config: \(msg)"
            case .networkError(let err): return "Network error: \(err.localizedDescription)"
            case .serverError(let code, let msg): return "Server error \(code): \(msg)"
            case .decodingError(let err): return "Decoding error: \(err.localizedDescription)"
            case .challengeTimeout: return "Challenge timed out"
            case .challengeFailed(let msg): return "Challenge failed: \(msg)"
            case .webViewUnavailable: return "WebView is not available"
            case .cancelled: return "CAPTCHA was cancelled"
            }
        }
    }

    // MARK: - Properties

    public let config: CaptchaConfig
    public let behaviorTracker: BehaviorTracker
    private let deviceCollector: DeviceInfoCollector
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    // MARK: - Init

    public init(config: CaptchaConfig) {
        self.config = config
        self.behaviorTracker = BehaviorTracker()
        self.deviceCollector = DeviceInfoCollector.shared

        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = 15
        sessionConfig.timeoutIntervalForResource = 30
        self.session = URLSession(configuration: sessionConfig)

        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    // MARK: - Execute

    /// Run the full CAPTCHA flow: precheck -> optional challenge -> result.
    @available(iOS 14.0, *)
    public func execute() async -> CaptchaResult {
        do {
            // Validate config
            guard !config.siteKey.isEmpty else {
                return .error(CaptchaError.invalidConfig("siteKey must not be empty"))
            }

            // Step 1: Collect signals
            let deviceInfo = deviceCollector.collect()
            let behavior = behaviorTracker.behaviorData()

            // Step 2: Precheck
            let precheckResult = try await precheck(deviceInfo: deviceInfo, behaviorData: behavior)

            // Step 3: Handle action
            switch precheckResult.action {
            case .pass:
                return .passed(token: precheckResult.token)

            case .invisible:
                // Invisible verification passed server-side
                return .passed(token: precheckResult.token)

            case .challenge:
                guard let challengeID = precheckResult.challengeID,
                      let challengeType = precheckResult.challengeType else {
                    return .error(CaptchaError.serverError(
                        statusCode: 0,
                        message: "Challenge required but no challenge_id or challenge_type returned"
                    ))
                }
                return .challengeRequired(challengeID: challengeID, type: challengeType)

            case .deny:
                return .denied
            }
        } catch {
            return .error(error)
        }
    }

    // MARK: - Verify

    /// Submit the user's challenge answer and get a verification result.
    @available(iOS 14.0, *)
    public func verify(challengeID: String, answer: String) async -> CaptchaResult {
        do {
            let verifyResponse = try await submitChallengeAnswer(
                challengeID: challengeID,
                answer: answer
            )

            if verifyResponse.success, let token = verifyResponse.token {
                return .passed(token: token)
            } else {
                let message = verifyResponse.errorMessage ?? "Verification failed"
                return .error(CaptchaError.challengeFailed(message))
            }
        } catch {
            return .error(error)
        }
    }

    // MARK: - Challenge WebView

    #if canImport(UIKit)
    /// Present a WKWebView to render the challenge and return the answer.
    /// Call this when `execute()` returns `.challengeRequired`.
    @MainActor
    @available(iOS 14.0, *)
    public func presentChallenge(
        challengeID: String,
        challengeType: String,
        from viewController: UIViewController
    ) async -> CaptchaResult {
        do {
            // Fetch render data
            let renderResponse = try await fetchChallengeRender(challengeID: challengeID)

            // Create and present WebView
            return await withCheckedContinuation { continuation in
                let webVC = ChallengeWebViewController(
                    renderURL: renderResponse.renderURL,
                    challengeID: challengeID,
                    config: config
                ) { [weak self] result in
                    guard let self = self else {
                        continuation.resume(returning: .error(CaptchaError.cancelled))
                        return
                    }
                    switch result {
                    case .answer(let answer):
                        Task {
                            let verifyResult = await self.verify(
                                challengeID: challengeID,
                                answer: answer
                            )
                            continuation.resume(returning: verifyResult)
                        }
                    case .cancelled:
                        continuation.resume(returning: .error(CaptchaError.cancelled))
                    case .error(let error):
                        continuation.resume(returning: .error(error))
                    }
                }
                viewController.present(webVC, animated: true)
            }
        } catch {
            return .error(error)
        }
    }
    #endif

    // MARK: - Network

    private func precheck(
        deviceInfo: [String: String],
        behaviorData: [String: String]
    ) async throws -> PrecheckResponse {
        let body = PrecheckRequest(
            siteKey: config.siteKey,
            scene: config.scene.rawValue,
            deviceInfo: deviceInfo,
            behaviorData: behaviorData,
            lang: config.lang,
            theme: config.theme.rawValue
        )
        return try await post(path: "/precheck", body: body)
    }

    private func fetchChallengeRender(challengeID: String) async throws -> ChallengeRenderResponse {
        let url = try buildURL(path: "/challenge/render", query: [
            "challenge_id": challengeID,
            "site_key": config.siteKey
        ])
        return try await get(url: url)
    }

    private func submitChallengeAnswer(
        challengeID: String,
        answer: String
    ) async throws -> ChallengeVerifyResponse {
        // Collect fresh device info from the current session
        let deviceInfo = deviceCollector.collect()
        let token = deviceInfo["fingerprint"] ?? ""

        let body = ChallengeVerifyRequest(
            challengeID: challengeID,
            answer: answer,
            token: token
        )
        return try await post(path: "/challenge/verify", body: body)
    }

    // MARK: - HTTP Helpers

    private func buildURL(path: String, query: [String: String]? = nil) throws -> URL {
        guard var components = URLComponents(string: config.apiBase + path) else {
            throw CaptchaError.invalidConfig("Invalid API base URL: \(config.apiBase)")
        }
        if let query = query {
            components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        guard let url = components.url else {
            throw CaptchaError.invalidConfig("Could not construct URL for path: \(path)")
        }
        return url
    }

    private func post<T: Encodable, R: Decodable>(path: String, body: T) async throws -> R {
        let url = try buildURL(path: path)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(DeviceInfoCollector.shared.collect()["user_agent"], forHTTPHeaderField: "User-Agent")
        request.httpBody = try encoder.encode(body)
        return try await perform(request: request)
    }

    private func get<R: Decodable>(url: URL) async throws -> R {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(DeviceInfoCollector.shared.collect()["user_agent"], forHTTPHeaderField: "User-Agent")
        return try await perform(request: request)
    }

    private func perform<R: Decodable>(request: URLRequest) async throws -> R {
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw CaptchaError.networkError(error)
        }

        guard let http = response as? HTTPURLResponse else {
            throw CaptchaError.serverError(statusCode: 0, message: "Non-HTTP response")
        }

        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? "no body"
            throw CaptchaError.serverError(statusCode: http.statusCode, message: body)
        }

        do {
            return try decoder.decode(R.self, from: data)
        } catch {
            throw CaptchaError.decodingError(error)
        }
    }
}

// MARK: - Challenge WebView Controller

#if canImport(UIKit)
/// Internal view controller that hosts a WKWebView to render CAPTCHA challenges.
@available(iOS 14.0, *)
internal final class ChallengeWebViewController: UIViewController, WKNavigationDelegate, WKScriptMessageHandler {

    enum ChallengeResult {
        case answer(String)
        case cancelled
        case error(Error)
    }

    private let renderURL: String
    private let challengeID: String
    private let config: CaptchaConfig
    private let completion: (ChallengeResult) -> Void
    private var webView: WKWebView!

    init(
        renderURL: String,
        challengeID: String,
        config: CaptchaConfig,
        completion: @escaping (ChallengeResult) -> Void
    ) {
        self.renderURL = renderURL
        self.challengeID = challengeID
        self.config = config
        self.completion = completion
        super.init(nibName: nil, bundle: nil)
        self.modalPresentationStyle = .pageSheet
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .systemBackground
        setupWebView()
        loadChallenge()
    }

    private func setupWebView() {
        let contentController = WKUserContentController()
        contentController.add(self, name: "captchaCallback")

        // Inject JS bridge so the challenge page can communicate back
        let bridgeScript = WKUserScript(
            source: """
            window.EngagelabCaptcha = {
                submit: function(answer) {
                    window.webkit.messageHandlers.captchaCallback.postMessage({
                        type: 'submit',
                        answer: answer
                    });
                },
                cancel: function() {
                    window.webkit.messageHandlers.captchaCallback.postMessage({
                        type: 'cancel'
                    });
                }
            };
            """,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: true
        )
        contentController.addUserScript(bridgeScript)

        let webConfig = WKWebViewConfiguration()
        webConfig.userContentController = contentController
        webConfig.allowsInlineMediaPlayback = true

        webView = WKWebView(frame: .zero, configuration: webConfig)
        webView.navigationDelegate = self
        webView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(webView)

        NSLayoutConstraint.activate([
            webView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            webView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            webView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            webView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])

        // Add close button
        let closeButton = UIButton(type: .close)
        closeButton.translatesAutoresizingMaskIntoConstraints = false
        closeButton.addTarget(self, action: #selector(closeTapped), for: .touchUpInside)
        view.addSubview(closeButton)

        NSLayoutConstraint.activate([
            closeButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 8),
            closeButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16)
        ])
    }

    private func loadChallenge() {
        // Build the render URL with query params
        var components = URLComponents(string: renderURL)
        var queryItems = components?.queryItems ?? []
        queryItems.append(URLQueryItem(name: "challenge_id", value: challengeID))
        queryItems.append(URLQueryItem(name: "theme", value: config.theme.rawValue))
        if let lang = config.lang {
            queryItems.append(URLQueryItem(name: "lang", value: lang))
        }
        queryItems.append(URLQueryItem(name: "platform", value: "ios"))
        components?.queryItems = queryItems

        if let url = components?.url {
            webView.load(URLRequest(url: url))
        } else {
            completion(.error(EngagelabCaptcha.CaptchaError.invalidConfig("Invalid render URL")))
            dismiss(animated: true)
        }
    }

    @objc private func closeTapped() {
        completion(.cancelled)
        dismiss(animated: true)
    }

    // MARK: - WKScriptMessageHandler

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let body = message.body as? [String: Any],
              let type = body["type"] as? String else { return }

        switch type {
        case "submit":
            if let answer = body["answer"] as? String {
                completion(.answer(answer))
                dismiss(animated: true)
            }
        case "cancel":
            completion(.cancelled)
            dismiss(animated: true)
        default:
            break
        }
    }

    // MARK: - WKNavigationDelegate

    func webView(
        _ webView: WKWebView,
        didFailProvisionalNavigation navigation: WKNavigation!,
        withError error: Error
    ) {
        completion(.error(error))
        dismiss(animated: true)
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        completion(.error(error))
        dismiss(animated: true)
    }
}
#endif
